solidity/contracts/bank/Bank.sol

```diff
@@ -0,0 +1,291 @@
+    ///         conforming EIP712 standard, rather than an on-chain transaction
+    ///         from their address. Anyone can submit this signature on the
+    ///         user's behalf by calling the permit function, paying gas fees,
+    ///         and possibly performing other actions in the same transaction.
+    /// @dev    The deadline argument can be set to `type(uint256).max to create
+    ///         permits that effectively never expire.  If the `amount` is set
+    ///         to `type(uint256).max` then `transferBalanceFrom` will not
+    ///         reduce an allowance.
+    function permit(
+        address owner,
+        address spender,
+        uint256 amount,
+        uint256 deadline,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) external {
+        /* solhint-disable-next-line not-rely-on-time */
+        require(deadline >= block.timestamp, "Permission expired");
+
+        // Validate `s` and `v` values for a malleability concern described in EIP2.
+        // Only signatures with `s` value in the lower half of the secp256k1
+        // curve's order and `v` value of 27 or 28 are considered valid.
+        require(
+            uint256(s) <=
+                0x7FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF5D576E7357A4501DDFE92F46681B20A0,
+            "Invalid signature 's' value"
+        );
+        require(v == 27 || v == 28, "Invalid signature 'v' value");
+
+        bytes32 digest =
+            keccak256(
+                abi.encodePacked(
+                    "\x19\x01",
+                    DOMAIN_SEPARATOR(),
+                    keccak256(
+                        abi.encode(
+                            PERMIT_TYPEHASH,
+                            owner,
+                            spender,
+                            amount,
+                            nonce[owner]++,
+                            deadline
+                        )
+                    )
+                )
+            );
+        address recoveredAddress = ecrecover(digest, v, r, s);
+        require(
+            recoveredAddress != address(0) && recoveredAddress == owner,
+            "Invalid signature"
+        );
+        _approveBalance(owner, spender, amount);
+    }
+
+    /// @notice Increases balances of the provided `recipients` by the provided
+    ///         `amounts`. Can only be called by the Bridge.
+    /// @dev This function fails if the lengths of the arrays are not the same.
+    function increaseBalances(
+        address[] calldata recipients,
+        uint256[] calldata amounts
+    ) external {
+        for (uint256 i = 0; i < recipients.length; i++) {
+            increaseBalance(recipients[i], amounts[i]);
+        }
+    }
+
+    /// @notice Decreases caller's balance by the provided `amount`. There is no
+    ///         way to restore the balance so do not call this function unless
+    ///         you really know what you are doing!
+    function decreaseBalance(uint256 amount) external {
```

Comment:
So here we will be able to force `msg.sender` to decrease their balance when this is needed, right?

Reply:
Correct. This function should be called by the Bridge most of the time. Initially, I gave the Bridge option to decrease anyone's balance but changed it in https://github.com/keep-network/tbtc-v2/pull/84/commits/cfc25677d26897ec7d5393a476dc728286da3830. I think this option is a bit safer. During the redemption, depositor will first transfer their Bitcoin balance to the Bridge, and once the Bitcoin is redeemed, Bridge will decrease its own balance. Of course, it does not fully protect against compromised Bridge because it can still increase balances but at least the decrease path is covered.